internal/argocd/revision.go

```diff
@@ -2,107 +2,224 @@ package argocd
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	kargoapi "github.com/akuity/kargo/api/v1alpha1"
 	argocd "github.com/akuity/kargo/internal/controller/argocd/api/v1alpha1"
 	"github.com/akuity/kargo/internal/controller/freight"
+	"github.com/akuity/kargo/internal/logging"
 )
 
+var NO_REVISIONS = []string{}
+
+func GetIntersection(revisions []string, desired []string) []string {
+	var intersection []string
+	hash := make(map[string]bool)
+	for _, e := range revisions {
+		hash[e] = true
+	}
+	for _, e := range desired {
+		// If elements present in the hashmap then append intersection list.
+		if hash[e] {
+			intersection = append(intersection, e)
+		}
+	}
+	return intersection
+}
```

Comment:
As noted elsewhere, you're returning an array of strings you never use apart from checking its length.

Apart from that, if you desire foo@x and bar@y and _have_ foo@y and bar@x, this just returns x, y, and the logic in the caller that looks at the length of this to see if it matches the length of desired revisions may end up concluding that everything is synced as desired when in fact, it is not.

Reply:
I'm not sure that's accurate. 

In `health.go` when checking sources, the single-source is explicitly matched, such that we expect the one revision to be matched:

```go
		singleSourceRevision := app.Status.Sync.Revision
		if singleSourceRevision != "" {
			matchingRevisions := GetIntersection([]string{singleSourceRevision}, revisions)

			if len(matchingRevisions) == 1 {
				return kargoapi.HealthStateHealthy, nil
			}
```

And for multiple sources, for all the revisions to be matched:
```go
		multiSourceRevisions := app.Status.Sync.Revisions
		if len(multiSourceRevisions) > 0 {
			// App has multiple sources, so we need to check the list of revisions.
			// Apps with multiple sources pointed at the same Git repository can only have the same revision
			// for all sources because ArgoCD does not support the alternative, so we only need to check
			// the first revision match.
			matchingRevisions := GetIntersection(multiSourceRevisions, revisions)

			if len(matchingRevisions) > 0 && len(matchingRevisions) == len(revisions) {
				logger.Debug("Found all desired revisions in list of revisions of multi-source application, app is healthy.",
					"desiredRevisions", revisions, "currentRevisions", multiSourceRevisions)
				return kargoapi.HealthStateHealthy, nil
			}
```

In both these cases, we also log what has been compared to what, so that an operator trying to diagnose will find the basis for the decisions in the logs.

Now, in `argocd.go`, this is focused on matching the number of desired revisions as well, which should achieve the same:
```go
	// Check multi-source app for matching revisions.
	if len(status.SyncResult.Revisions) > 0 {
		matchingRevisions := libargocd.GetIntersection(status.SyncResult.Revisions, desiredRevisions)
		if len(matchingRevisions) == len(desiredRevisions) {
			appLogger.Debug("Operation completed, no update necessary.", "phase", status.Phase)
			return status.Phase, false, nil
		}

		return "", true, fmt.Errorf(
			"Not all result revisions out of %v match desired revisions %v",
			status.SyncResult.Revisions, desiredRevisions)
	}

	// Check single-source app for matching revision.
	singleSourceRevision := status.SyncResult.Revision
	if singleSourceRevision != "" && len(desiredRevisions) > 0 {
		matchingRevisions := libargocd.GetIntersection([]string{singleSourceRevision}, desiredRevisions)

		if len(matchingRevisions) != 1 {
			// The operation did not result in the desired revision being applied.
			// We should attempt to retry the operation.
			return "", true, fmt.Errorf(
				"operation result revision %q does not match desired revision %v",
				status.SyncResult.Revision, desiredRevisions,
			)
		}
	}
```

Thoughts?

Reply:
The two string slices passed to `GetIntersection` are _just_ version numbers, lacking any other context.

There's what you have and there's what you want.

Let us presume you _have_ `["1.0.0", "1.1.0"]`. Let us further presume the missing context here is that you have _foo_@1.0.0 and _bar_@1.1.0.

Now say we _desire_ foo@1.1.0 and have no opinion about what bar should be. But again, this is furnished without context to your function. It receives `["1.1.0"]`

The logic in your `GetIntersection` algorithm finds the intersection of `["1.0.0", "1.1.0"]` and `["1.1.0"]`, which is `["1.1.0"]`. By the following logic, you are considering all sources to be at their desired revision, but they are not:

```
if len(matchingRevisions) > 0 && len(matchingRevisions) == len(revisions) {
				logger.Debug("Found all desired revisions in list of revisions of multi-source application, app is healthy.",
					"desiredRevisions", revisions, "currentRevisions", multiSourceRevisions)
				return kargoapi.HealthStateHealthy, nil
			}
```

Reply:
I see your point. I will see if I can refactor the solution to more accurately represent desired state.

Reply:
If we wanted to stay aligned with the ArgoCD implementation, we could align by using a shadow array for the desired revisions where:

- If we have a source update, we would return the version appropriate based on freight
- Let missing chart and similar errors bubble up and ultimately treat the app as Unhealthy
- Return source target revision for sources we don't want to update ie. don't have a targetUpdate for

Thoughts?

Reply:
Ya. I think you are on the right track.

I don't _like_ all this parallel array business. It's brittle, but it's what Argo CD has left us to work with.

If I understand your last suggestion correctly, it would be something like this:

1. Iterate over all of the App's sources, finding desired revision for each. If desired revision cannot be determined, the corresponding entry in the list of desired revisions will be the empty string.
2. Iterate over the observed revision of each source, which as I understand it, will be ordered the same as the sourced themselves. (Again... super brittle. 🤦‍♂️) If the corresponding desired revision is either the empty string _or_ matches the observed revision, then we're good. And discrepancy where desired revision wasn't the empty string means the App isn't synced correctly.

Sound right?

Reply:
Hey @krancour , sorry for the delay. I modified the code to mimic shadow-array use in ArgoCD and added some tests to handle multi-source apps accordingly, let me know what you think.